rust/src/main.rs

```diff
@@ -76,37 +109,147 @@ fn setup_logging() -> Result<(), fern::InitError> {
+    };
+    if logfile_file_name.eq(Path::new("")) {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "A blank file name was supplied",
+        ));
+    };
+    // Use the directory name set in logfile_config or fallback to the default
+    let logfile_directory_name = match logfile_config.directory {
+        Some(dir) => dir,
+        None => PathBuf::from(XI_LOG_DIR),
+    };
+
+    let mut logging_directory_path = get_logging_directory_path(logfile_directory_name)?;
+
+    // Add the file name & return the full path
+    logging_directory_path.push(logfile_file_name);
+    Ok(logging_directory_path)
+}
+
+fn get_flags() -> HashMap<String, Option<String>> {
+    let mut flags: HashMap<String, Option<String>> = HashMap::new();
+
+    let flag_prefix = "-";
+    let mut args_iterator = std::env::args().peekable();
+    while let Some(arg) = args_iterator.next() {
+        if arg.starts_with(flag_prefix) {
+            let key = arg.trim_left_matches(flag_prefix).to_string();
+
+            // Check the next argument doesn't start with the flag prefix
+            // map_or accounts for peek returning an Option
+            let next_arg_not_a_flag: bool = args_iterator
+                .peek()
+                .map_or(false, |val| !val.starts_with(flag_prefix));
+            if next_arg_not_a_flag {
+                flags.insert(key, args_iterator.next());
+            }
+        }
+    }
+    flags
+}
+
+struct EnvFlagConfig {
+    env_name: &'static str,
+    flag_name: &'static str,
+}
+
+/// Extracts a value from the flags and the env.
+///
+/// In this order: `String` from the flags, then `String` from the env, then `None`
+fn extract_env_or_flag(
+    flags: &HashMap<String, Option<String>>,
+    conf: EnvFlagConfig,
+) -> Option<String> {
+    flags
+        .get(conf.flag_name)
+        .cloned()
+        .unwrap_or(std::env::var(conf.env_name).ok())
+}
+
+struct LogfileConfig {
+    directory: Option<PathBuf>,
+    file: Option<PathBuf>,
+}
+
+fn generate_logfile_config(flags: &HashMap<String, Option<String>>) -> LogfileConfig {
+    // If the key is set, get the Option within
+    let log_dir_env_flag = EnvFlagConfig {
+        env_name: "XI_LOG_DIR",
+        flag_name: "log-dir",
+    };
+    let log_file_env_flag = EnvFlagConfig {
+        env_name: "XI_LOG_FILE",
+        flag_name: "log-file",
+    };
+    let log_dir_flag_option = extract_env_or_flag(&flags, log_dir_env_flag).map(PathBuf::from);
+
+    let log_file_flag_option = extract_env_or_flag(&flags, log_file_env_flag).map(PathBuf::from);
+
+    LogfileConfig {
+        directory: log_dir_flag_option,
+        file: log_file_flag_option,
+    }
+}
+
 fn main() {
     let mut state = XiCore::new();
     let stdin = io::stdin();
     let stdout = io::stdout();
     let mut rpc_looper = RpcLoop::new(stdout);
 
-    if let Err(e) = setup_logging() {
+    let flags: HashMap<String, Option<String>> = get_flags();
```

Comment:
you can get away without all of these type declarations I suspect.

Reply:
I'm happy to remove them if you prefer. It was mostly for while I was tweaking things.